### ISSUE TITLE
RA/Wayland : vertical frequency applied internally but no screen mode switch

### ISSUE THREAD

**Issue by bidinou** [libretro/RetroArch#13288]
Hi ! Trying to minimize latency as much as possible (for core which are not optimal, like puae, as it doesn't support runahead), I was suggested to switch to Wayland + Vulkan and using a single buffering mode with no VSync.

For some reason, there is no tearing when using single buffering under Wayland. Which means, 20 ms won on 50 Hz cores ! 

This issue is the following : 

- when a core override is set, indicating a custom vertical frequency, it is applied by RA (internal rendering), however it doesn't trigger a monitor mode switch (which remains at the desktop frequency)

however, under X, a custom freq. is both applied to RA and triggers a mode switch.

Configuration is : latest SNAP packages, KDE Neon (Ubuntu 20.04 based), Vulkan, AMD GPU, Plasma/Wayland.

**Comment by i30817** [comment]
_I really doubt_ wayland allows you to change your refresh rate, just like it doesn't allow you to change your desktop resolution.

**Comment by nfp0** [comment]
> I was suggested to switch to Wayland + Vulkan and using a single buffering mode with no VSync

@bidinou That is not possible on Wayland for now. It's a work in progress. Who suggested that to you?
I did these tests a while back if you're interested in the best modes for reduced latency:
https://forums.libretro.com/t/an-input-lag-investigation/4407/1006?u=nfp0

Long story short, on Linux, KMS, Uncomposited X and Kwin Wayland on AMDVLK provide you with the smallest possible latency (don't know about other desktops).
The problem is that I have a lot of issues with RetroArch on AMDVLK on my machine. Give it a try on your system and if all else fails, you got KMS.

**Comment by i30817** [comment]
KMS has a lot of other problems in certain hardware or cores that want to change resolutions. 

The problems i encountered were:

1. retroarch eats the VT switch key combinations, so if for any reason retroarch locks up, you're forced to 'power reset' (or kernel software reset, if you can remember the key combinations for you OS/kernel/keyboard), because you can't switch the virtual terminal away from the frozen retroarch. This can damage hard-drives if you don't do the kernel reset.
2. A particular important hardware quirks library (libinput) only starts up once you're in a graphical environment. So if you have 'naughty' hardware, direct access is often a bad idea.
3. And finally what made me give up on kms: the kernel has a lock mechanism to have exclusive access to the graphics card (the so called '[drm master](https://dri.freedesktop.org/docs/drm/gpu/drm-uapi.html)'), that can be 'normally' be called without root if you're on a VT, because switching to a VT dropped the lock from, for instance, X11 or wayland. Since RA doesn't allow you to switch VTs while running in KMS (see 1), this is not a problem. The problems occur when you have to mode switch because a core wants to change screen resolution from the natural one the menu starts with. This apparently always fails, even if it's rare to happen in emulators (for instance, you can try to change the internal resolution multiplier in beetle.
I think this happens because a resolution change triggers a gpu mode change and 'something' has to happen, not sure if related to the 'drm master'. Anyway, it crashes retroarch.

**Comment by nfp0** [comment]
Yeah, KMS has its own can of worms, no doubt. I've also avoided using it because of some issues with cores. I also noticed the VT changing problems you mention. I worked around it by SSHing from my phone and rebooting the PC, but yeah... not practical at all.
I did not know Lakka ran RetroArch as root. That explains how they can work around some of the issues. Thanks for the explanation! :slightly_smiling_face:

**Comment by i30817** [comment]
What i didn't mention is that for problem number 1 are _really_ easy way to trigger it is to try to search a playlist/menu in KMS. The screen comes up, but without a keyboard focus (or something like that) so you can dismiss it, and so can't navigate the screens to exit retroarch. Frozen.

**Comment by nfp0** [comment]
Yeah, that would be a pretty easy situation to get into a lock.

`You can reset a 'frozen' linux computer without ssh with the kernel Alt+SysRq REISUB mantra.`
Ah yes, thank you! @i30817 I had forgotten about it.

**Comment by i30817** [comment]
I suppose that there might be good news about on the quirks hardware front, because last time i updated ubuntu 22.04.1 it decided to install 'systemd-hwe-hwdb' which is a pretty telling name. Maybe the quirks were moved to a library to start at boot.
